File: iosApp/iosApp/Modules/ImprovementProposals/ImprovementProposalsWireframe.swift
// Created by web3d3v on 30/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultImprovementProposalsWireframe {
    private weak var parent: UIViewController?
    private let improvementProposalWireframeFactory: ImprovementProposalWireframeFactory
    private let improvementProposalsService: ImprovementProposalsService
    private weak var vc: UIViewController?

    init(
        _ parent: UIViewController?,
        improvementProposalWireframeFactory: ImprovementProposalWireframeFactory,
        improvementProposalsService: ImprovementProposalsService
    ) {
        self.parent = parent
        self.improvementProposalWireframeFactory = improvementProposalWireframeFactory
        self.improvementProposalsService = improvementProposalsService
    }
}

extension DefaultImprovementProposalsWireframe: ImprovementProposalsWireframe {

    func present() {
        let vc = wireUp()
        self.vc = vc
        parent?.show(vc, sender: self)
    }

    func navigate(destination: ImprovementProposalsWireframeDestination) {
        if let vote = destination as? ImprovementProposalsWireframeDestination.Vote {
            FeatureShareHelper().shareVote(on: vote.proposal)
        }
        if let dest = destination as? ImprovementProposalsWireframeDestination.Proposal {
            improvementProposalWireframeFactory.make(
                vc,
                context: .init(proposal: dest.proposal)
            ).present()
        }
    }
    
    func dismiss() { parent?.popOrDismiss() }
}

extension DefaultImprovementProposalsWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultImprovementProposalsInteractor(
            improvementProposalsService: improvementProposalsService
        )
        let storyboard = UIStoryboard(.improvementProposals)
        let vc: ImprovementProposalsViewController = storyboard.instantiate()
        let presenter = DefaultImprovementProposalsPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor
        )
        vc.presenter = presenter
        return NavigationController(rootViewController: vc)
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/improvementProposals/ImprovementProposalsWireframe.kt
package com.sonsofcrypto.web3walletcore.modules.improvementProposals

import com.sonsofcrypto.web3walletcore.services.improvementProposals.ImprovementProposal

sealed class ImprovementProposalsWireframeDestination {
    /** Votes on proposal */
    data class Vote(
        val proposal: ImprovementProposal
    ): ImprovementProposalsWireframeDestination()
    /** Navigate to proposal detail */
    data class Proposal(
        val proposal: ImprovementProposal,
    ): ImprovementProposalsWireframeDestination()
    /** Dismiss proposals wireframe */
    object Dismiss: ImprovementProposalsWireframeDestination()
}

interface ImprovementProposalsWireframe {
    /** Present module */
    fun present()
    /** Navigate to a new destination module */
    fun navigate(destination: ImprovementProposalsWireframeDestination)
}
